core/coreapi/object.go

```diff
@@ -159,6 +161,108 @@ func (api *ObjectAPI) Data(ctx context.Context, path ipath.Path, unixfs bool, me
 	}
 }
 
+// Return user data, metadata in byte array, and error.
+// Note that this function assumes, if token metadata exists inside
+// the DAG topped by the given 'node', the 'node' has metadata root
+// as first child, user data root as second child.
+func getDataForUserAndMeta(ctx context.Context, nd ipld.Node, ds ipld.DAGService) ([]byte, []byte, error) {
+	n := nd.(*dag.ProtoNode)
+
+	fsType, err := ft.GetFSType(n)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if ft.TTokenMeta == fsType {
+		return nil, n.Data(), nil
+	}
+
+	// Return user data and metadata if first child is of type TTokenMeta.
+	if nd.Links() != nil && len(nd.Links()) >= 2 {
+		childen, err := ft.GetChildrenForDagWithMeta(ctx, nd, ds)
+		if err != nil {
+			return nil, nil, err
+		}
+		if childen == nil {
+			return nil, n.Data(), nil
+		}
+		return childen[1].(*dag.ProtoNode).Data(), childen[0].(*dag.ProtoNode).Data(), nil
+	}
+
+	return n.Data(), nil, nil
+}
+
+func (api *ObjectAPI) MetaDataMap(ctx context.Context, path ipath.Path) (map[string]interface{}, error) {
+	mr, err := api.MetaData(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+
+	metaData, err := ioutil.ReadAll(mr)
```

Comment:
There is a small efficiency issue here. In `api.Metadata`, `[]byte` is wrapped in a reader, and here it immediately reads it out again as `[]byte`. You can shorten the extra step it would be faster.